src/track/beats.cpp

```diff
@@ -32,25 +541,474 @@ double Beats::findNBeatsFromSample(double fromSample, double beats) const {
+    markerToInsert.mutable_time_signature()->set_note_value(sig.getNoteValue());
+    QVector<track::io::TimeSignatureMarker> timeSignatureMarkersMutableCopy;
+    copy(m_beatsProto.time_signature_markers().cbegin(),
+            m_beatsProto.time_signature_markers().cend(),
+            std::back_inserter(timeSignatureMarkersMutableCopy));
+
+    track::io::TimeSignatureMarker searchBeforeMarker;
+    searchBeforeMarker.set_downbeat_index(downbeatIndex);
+    auto prevTimeSignatureMarker =
+            std::lower_bound(timeSignatureMarkersMutableCopy.begin(),
+                    timeSignatureMarkersMutableCopy.end(),
+                    searchBeforeMarker,
+                    TimeSignatureMarkerEarlier);
+    if (prevTimeSignatureMarker->downbeat_index() == downbeatIndex) {
+        prevTimeSignatureMarker->CopyFrom(markerToInsert);
+    } else {
+        timeSignatureMarkersMutableCopy.insert(
+                prevTimeSignatureMarker, markerToInsert);
+    }
+    m_beatsProto.clear_time_signature_markers();
+    for (const auto& timeSignatureMarker : timeSignatureMarkersMutableCopy) {
+        m_beatsProto.add_time_signature_markers()->CopyFrom(
+                timeSignatureMarker);
+    }
+    generateBeatsFromMarkers();
+}
+
+void BeatsInternal::translate(FrameDiff_t numFrames) {
+    if (!isValid()) {
+        return;
+    }
+
+    m_beatsProto.set_first_frame_position(
+            m_beatsProto.first_frame_position() + numFrames);
+    generateBeatsFromMarkers();
+}
+
+void BeatsInternal::setBpm(Bpm bpm, int beatIndex) {
+    // TODO(hacksdump): A check for preferences will be added to only allow setting bpm
+    //  when "Assume Constant Tempo" is checked.
+    if (!isValid()) {
+        return;
+    }
+    track::io::BpmMarker markerToInsert;
+    markerToInsert.set_beat_index(beatIndex);
+    markerToInsert.set_bpm(bpm.getValue());
+    QVector<track::io::BpmMarker> bpmMarkersMutableCopy;
+    copy(m_beatsProto.bpm_markers().cbegin(),
+            m_beatsProto.bpm_markers().cend(),
+            std::back_inserter(bpmMarkersMutableCopy));
+
+    track::io::BpmMarker searchBeforeMarker;
+    searchBeforeMarker.set_beat_index(beatIndex);
+    auto prevBpmMarker =
+            std::lower_bound(bpmMarkersMutableCopy.begin(),
+                    bpmMarkersMutableCopy.end(),
+                    searchBeforeMarker,
+                    BpmMarkerEarlier);
+    if (prevBpmMarker->beat_index() == beatIndex) {
+        prevBpmMarker->CopyFrom(markerToInsert);
+    } else {
+        bpmMarkersMutableCopy.insert(
+                prevBpmMarker, markerToInsert);
+    }
+    m_beatsProto.clear_bpm_markers();
+    for (const auto& bpmMarker : bpmMarkersMutableCopy) {
+        m_beatsProto.add_bpm_markers()->CopyFrom(
+                bpmMarker);
+    }
+    generateBeatsFromMarkers();
+    QList<int> indices;
+    for (const auto& bpmMarker : m_beatsProto.bpm_markers()) {
+        indices.append(bpmMarker.beat_index());
+    }
```

Comment:
Did you intend to do something with this `indices` local variable? Currently it does nothing.